### PR TITLE
JP-3455: Updated MIRI imager photom step

### LINKED ISSUE #8048: Updates to PHOTOM step and files for MIRI imager time-dependent throughput
_Issue [JP-3455](https://jira.stsci.edu/browse/JP-3455) was created on JIRA by [Misty Cracraft](https://jira.stsci.edu/secure/ViewProfile.jspa?name=cracraft):_

A time-dependent throughput issue was found for the MIRI Imager, as had previously been found for the MIRI MRS. As with the fix for the MIRI MRS, we now have updated code that can go into the pipeline to use updated photom files with a correction based on MJD of the observation taken from the header and will not rely on a set of USEAFTER date specified photom files like we currently have in place for the workaround.

[Mattia Libralato](https://jira.stsci.edu/secure/ViewProfile.jspa?name=libra)  has developed the code and created the appropriate photom reference file for this.

### PR DESCRIPTION
**Pull request by mlibralato**
Resolves [JP-3455](https://jira.stsci.edu/browse/JP-3455)

Closes #8048 

This PR is related to proposed changes to the PHOTOM step for the MIRI Imager (and coronographs). We add a time-dependent correction to the PHOTOM value.

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [ ] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by mlibralato** [comment]
@hbushouse Let me know if everything works now.

File: jwst/photom/photom.py [hunk 1]
```diff
@@ -468,7 +470,32 @@ def calc_miri(self, ftab):
...
+                    dtype=[
+                        ("filter", "O"),
+                        ("subarray", "O"),
+                        ("photmjsr", "<f4"),
+                        ("uncertainty", "<f4")
+                        ],
+                )
+                fftab = MirImgPhotomModel(phot_table=data)
+                self.photom_io(fftab.phot_table[0])
+            except:
```
Comment (nden) [inline jwst/photom/photom.py:494]: The style error in the tests comes from the "bare" except statement. It looks like it is meant to catch the case where the "timecoeff" extension is not in the reference file. Would this be an `AttributeError` then?

When I try running `ftab.getarray_noinit("timecoeff")` on an old reference file an `AttributeError` is raised.

Comment (mlibralato) [reply jwst/photom/photom.py:494]: Should I write the following line instead of just "except"?

`except AttributeError as e:`

Comment (mlibralato) [reply jwst/photom/photom.py]: Done

**Comment by hbushouse** [comment]
All of the CI failures are due to errors in the jwst/photom/tests/test_photom.py unit test module when it tries to access the new "timecoeff" attribute in the ref file model, but it's not defined here, due to the fact that the CI test environment is still using the latest stdatamodels release, rather than stdatamodels/master.

@mlibralato Can you try to verify that running "pytest test_photom.py" in your local branch, which presumably has the latest stdatamodels updates, succeeds?

File: jwst/photom/miri_imager.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    Parameters
+    ----------
+    param : numpy array
+            Set of parameters for the PHOTOM value
+    t : int
+        Modified Julian Day (MJD) of the observation
+
+    Returns
+    -------
+    The time-dependent correction to the photmjsr term.
```
Comment (hbushouse) [inline jwst/photom/miri_imager.py:28]:
```suggestion
    corr : float
        The time-dependent correction to the photmjsr term.
```

File: jwst/photom/photom.py [hunk 3]
```diff
@@ -468,7 +470,32 @@ def calc_miri(self, ftab):
...
+                data = np.array(
+                    [(self.filter, self.subarray, ftab.phot_table[row]['photmjsr']+photom_corr, ftab.phot_table[row]['uncertainty'])],
+                    dtype=[
+                        ("filter", "O"),
+                        ("subarray", "O"),
+                        ("photmjsr", "<f4"),
+                        ("uncertainty", "<f4")
+                        ],
+                )
+                fftab = MirImgPhotomModel(phot_table=data)
```
Comment (hbushouse) [inline jwst/photom/photom.py:492]: Suggest changing this to `fftab = datamodels.MirImgPhotomModel(phot_table=data)` and then don't import `MirImgPhotomModel` at the beginning of the module.

**Comment by hbushouse** [comment]
Also, I still can't run the full regression test suite on this PR branch due to some kind of password protection or permission issue with your fork of the repo. So once you've made the suggested changes on your local PR branch, can you at least run a few of the regression test modules manually, just so we have some kind of assurance that this works OK? To do so, go to the /jwst/jwst/regtest directory in your local branch and execute the MIRI imager tests using:

`pytest --bigdata test_miri_image.py`

Note that you have to first setup the environment variable `export TEST_BIGDATA=https://bytesalad.stsci.edu/artifactory` so that Jenkins knows where to get the input data files used by the tests.

**Comment by mlibralato** [comment]
> All of the CI failures are due to errors in the jwst/photom/tests/test_photom.py unit test module when it tries to access the new "timecoeff" attribute in the ref file model, but it's not defined here, due to the fact that the CI test environment is still using the latest stdatamodels release, rather than stdatamodels/master.
> 
> @mlibralato Can you try to verify that running "pytest test_photom.py" in your local branch, which presumably has the latest stdatamodels updates, succeeds?

I tested the code before pushing it and with the updated stdatamodels I get a "pass".

**Comment by hbushouse** [comment]
> > Also, I still can't run the full regression test suite on this PR branch due to some kind of password protection or permission issue with your fork of the repo. So once you've made the suggested changes on your local PR branch, can you at least run a few of the regression test modules manually, just so we have some kind of assurance that this works OK? To do so, go to the /jwst/jwst/regtest directory in your local branch and execute the MIRI imager tests using:
> > `pytest --bigdata test_miri_image.py`
> > Note that you have to first setup the environment variable `export TEST_BIGDATA=https://bytesalad.stsci.edu/artifactory` so that Jenkins knows where to get the input data files used by the tests.
> 
> The test failed but I wonder if all errors are due to the change in the PHOTOM step itself.

That's probably the case. I noticed that the differences start with the "cal" product and proceed downstream from there, which is exactly what you'd expect from changes in the photom correction (which is the last step before "cal" is created).

**Comment by mlibralato** [comment]
> > > Also, I still can't run the full regression test suite on this PR branch due to some kind of password protection or permission issue with your fork of the repo. So once you've made the suggested changes on your local PR branch, can you at least run a few of the regression test modules manually, just so we have some kind of assurance that this works OK? To do so, go to the /jwst/jwst/regtest directory in your local branch and execute the MIRI imager tests using:
> > > `pytest --bigdata test_miri_image.py`
> > > Note that you have to first setup the environment variable `export TEST_BIGDATA=https://bytesalad.stsci.edu/artifactory` so that Jenkins knows where to get the input data files used by the tests.
> > 
> > 
> > The test failed but I wonder if all errors are due to the change in the PHOTOM step itself.
> 
> That's probably the case. I noticed that the differences start with the "cal" product and proceed downstream from there, which is exactly what you'd expect from changes in the photom correction (which is the last step before "cal" is created).

So, what should I do? Just ignore it for now? Have you solved the issue with the permission? If you show me the error, I can try to see on my end what I can do

**Review by hbushouse — APPROVED** [review]
All of the CI failures are due to the use of the last released version of stdatamodels, which doesn't have the updates to the MIRI imager photom data model. Note how the one using "devdeps" passes, because it's using the dev version of stdatamodels. So this all looks OK to me now.

**Pull request merged**